File: ts/breaker.ts
/**
 * Dispatch-loop circuit breaker — pure policy: no I/O, no timers, no clock.
 *
 * Extracted from a production LLM dispatch loop's breaker module (names and
 * signatures unchanged; only the doc comments were generalized away from the
 * source repo's internal rule numbers and cross-file references). Classifies
 * a failure message into a systemic class, computes capped-exponential
 * backoff delays, and tracks a batch's consecutive-systemic-failure streak
 * to decide trip / dead-letter / completed.
 */

import { requireFields } from "./contract.ts";
import { verdict, type Verdict } from "./verdict.ts";

/** 2^53 - 1: the largest integer every language holds, and so the breaker's. */
const WHOLE_MAX = 9007199254740991;

/** An argument that must be an integer in range. Outside the contract it
 * throws: the call fails and changes nothing. Every integer the breaker takes
 * is one a number holds, so a bigint is no argument of its - unlike a
 * budget's, whose ledger reaches 2^63 - 1. */
function whole(value: unknown, what: string): number {
  // isInteger is false for whatever is not a number, so it is the type test too.
  if (!Number.isInteger(value)) throw new TypeError(`${what} must be an integer, got ${String(value)}`);
  const n = value as number;
  if (n < -WHOLE_MAX || n > WHOLE_MAX) {
    throw new RangeError(`${what} must be within +/-(2^53 - 1), got ${n}`);
  }
  return n;
}

/** An argument that must be a string. */
function text(value: unknown, what: string): string {
  if (typeof value !== "string") throw new TypeError(`${what} must be a string, got ${String(value)}`);
  return value;
}

/** An argument that must be a boolean. */
function flag(value: unknown, what: string): boolean {
  if (typeof value !== "boolean") throw new TypeError(`${what} must be a boolean, got ${String(value)}`);
  return value;
}

/** A flag that may be absent - null or not given - and is then off. */
function optionalFlag(value: unknown, what: string): boolean {
  return value == null ? false : flag(value, what);
}

/**
 * Systemic failure class a dispatch failure message can be classified into.
 * `null` (returned by the classifier, not a member of this type) means the
 * failure is item-local and must never count toward the breaker.
 */
export type SystemicDispatchFailureClass = "rate_limit" | "auth" | "transport";

const RATE_LIMIT_PATTERNS = [
  "429",
  "rate limit",
  "limit reached",
  "rate_limit",
  "too many requests",
  "overloaded",
  // Provider capacity refusal — same systemic class as "overloaded": the
  // provider is shedding load, not the item failing. Anchored to the full
  // phrase: a bare "at capacity" can occur in unrelated domain text, and an
  // item-local contract failure must never become breaker fuel.
  "selected model is at capacity",
  "session limit",
  "usage limit",
  "quota",
  "retry-after",
  "retry_after",
];

const AUTH_PATTERNS = [
  "401",
  "403",
  "unauthorized",
  "forbidden",
  "invalid api key",
  "invalid x-api-key",
  "authentication",
  "not logged in",
];

/** Single source for the transient-transport message patterns, kept as its
 * own list because other per-unit retry decisions outside this policy may
 * want the same substrings without the rest of TRANSPORT_PATTERNS. */
export const TRANSIENT_TRANSPORT_MESSAGE_PATTERNS = [
  "stream disconnected before completion",
  "connection reset by peer",
  "error sending request",
  "failed to connect to websocket",
  "transport channel closed",
  "http/request failed",
  "request failed after",
] as const;

const TRANSPORT_PATTERNS = [
  ...TRANSIENT_TRANSPORT_MESSAGE_PATTERNS,
  "timed out",
  "timeout",
  "econnrefused",
  "econnreset",
  "etimedout",
  "socket hang up",
  "fetch failed",
];

/**
 * Classify a failure message into a systemic dispatch class, or null for
 * item-local failures (malformed output, validation rejection, …) that must
 * never trip the batch breaker. Message-based by necessity: providers and
 * adapters commonly flatten their status into plain strings.
 */
export function classifySystemicDispatchFailure(
  message: string | null | undefined,
): SystemicDispatchFailureClass | null {
  // Absent - null or not given - is no failure text to read. An empty message
  // needs no test of its own: it holds no pattern.
  if (message == null) return null;
  const normalized = text(message, "a failure message").toLowerCase();
  if (RATE_LIMIT_PATTERNS.some((pattern) => normalized.includes(pattern))) {
    return "rate_limit";
  }
  if (AUTH_PATTERNS.some((pattern) => normalized.includes(pattern))) {
    return "auth";
  }
  if (TRANSPORT_PATTERNS.some((pattern) => normalized.includes(pattern))) {
    return "transport";
  }
  return null;
}

/** Capped exponential backoff (no jitter — deterministic for replay/tests).
 * attempt is 0-based: delay before retry #1 is initialMs. */
export function dispatchBackoffDelayMs(args: {
  attempt: number;
  initial_ms: number;
  cap_ms: number;
}): number {
  requireFields(args, BACKOFF_FIELDS, BACKOFF_FIELDS, "backoff arguments");
  const attempt = whole(args.attempt, "attempt");
  const initialMs = whole(args.initial_ms, "initial_ms");
  const capMs = whole(args.cap_ms, "cap_ms");
  const exponential = initialMs * 2 ** Math.max(0, attempt);
  const bounded = Math.min(capMs, exponential);
  return Number.isFinite(bounded) && bounded > 0 ? Math.floor(bounded) : capMs;
}

const BACKOFF_FIELDS = ["attempt", "initial_ms", "cap_ms"];

export interface DispatchBreakerPolicy {
  enabled: boolean;
  /** N: consecutive distinct-item systemic FINAL failures that trip the breaker. */
  systemic_threshold: number;
  /** Per-CALL total attempt cap (1 original + backoff retries) for
   * systemic-class failures. Breaker counting is per ITEM (observation);
   * backoff is per call. */
  per_call_max_attempts: number;
  backoff_initial_ms: number;
  backoff_cap_ms: number;
  /** Concurrent-pool mode (opt-in; default off/undefined preserves sequential
   * behavior byte-for-byte). When on, a pre-trip success does NOT flush the
   * pending systemic failures to poison, so the trip and the
   * completed/dead-letter/incomplete classification depend only on WHICH
   * items failed systemically, not on their completion order. Sequential
   * callers omit it and keep the poison-vs-systemic-via-later-success
   * attribution. */
  concurrent?: boolean | null;

}

/** `concurrent` is the one field a policy may leave out; it is then off. */
const POLICY_REQUIRED = [
  "enabled",
  "systemic_threshold",
  "per_call_max_attempts",
  "backoff_initial_ms",
  "backoff_cap_ms",
];
const POLICY_FIELDS = [...POLICY_REQUIRED, "concurrent"];
const ENTRY_FIELDS = ["item_id", "failure_class", "failure_message", "attempt_count"];

export interface DispatchDeadLetterEntry {
  item_id: string;
  /** Null is the item's own failure. The part does not judge what a class
   * says: any other string is systemic. */
  failure_class: SystemicDispatchFailureClass | null;
  failure_message: string;
  attempt_count: number;
}


/** The batch's trip, once it has one: a warning verdict, and the three facts its
 * reason names. `resume` is null, because what a next run picks up is the
 * pending entries and not a place. */
export interface DispatchBreakerTripState extends Verdict {
  failure_class: SystemicDispatchFailureClass;
  consecutive_item_count: number;
  threshold: number;
}

/**
 * Breaker state machine over one batch. The loop reports each item's FINAL
 * outcome (after its bounded retries); the machine answers "has this become
 * systemic?" and keeps the dead-letter/completion bookkeeping the caller
 * persists.
 *
 * Poison-vs-systemic attribution rule: a systemic-class failure is held
 * PENDING until the batch proves the provider lane is alive (a later item
 * succeeds) — only then is it a poison item (reproduced on that item alone)
 * and dead-lettered. If the streak instead reaches the threshold, the batch
 * trips and the pending items stay out of the dead-letter set: they were
 * victims of the outage and must be re-dispatched on recovery, not
 * complete-with-failure. Item-local failures (null class) dead-letter
 * immediately and say nothing about the provider, so they neither extend nor
 * reset the systemic streak.
 */
export class DispatchBreakerState {
  readonly policy: DispatchBreakerPolicy;
  /** The policy as read: what the batch decides by, once and not again. */
  private readonly enabled: boolean;
  private readonly threshold: number;
  private readonly concurrent: boolean;
  private pendingSystemic: DispatchDeadLetterEntry[] = [];
  private trip: DispatchBreakerTripState | null = null;
  private readonly completed: string[] = [];
  private readonly deadLetter: DispatchDeadLetterEntry[] = [];

  constructor(policy: DispatchBreakerPolicy) {
    requireFields(policy, POLICY_FIELDS, POLICY_REQUIRED, "breaker policy");
    this.enabled = flag(policy.enabled, "enabled");
    this.concurrent = optionalFlag(policy.concurrent, "concurrent");
    this.threshold = whole(policy.systemic_threshold, "systemic_threshold");
    if (this.threshold < 1) {
      throw new RangeError(`systemic_threshold must be at least 1, got ${this.threshold}`);
    }
    // Carried for the caller's loop and read by nothing here - and in the
    // contract all the same, so that a typo cannot pass as a policy.
    whole(policy.per_call_max_attempts, "per_call_max_attempts");
    whole(policy.backoff_initial_ms, "backoff_initial_ms");
    whole(policy.backoff_cap_ms, "backoff_cap_ms");
    this.policy = policy;
  }

  /** Report a REAL dispatch success — the only event that proves the
   * provider lane is alive and may reclassify pending systemic failures as
   * poison. Items that made no successful provider call must use
   * {@link recordItemSkipped} instead. */
  recordItemSuccess(itemId: string): void {
    this.completed.push(text(itemId, "item_id"));
    // Attribution freezes at trip: a CONCURRENT pool can deliver an in-flight
    // success after the trip decision, and letting it reclassify the pending
    // outage victims as poison would dead-letter them out of the incomplete
    // recovery set. The late unit itself still counts as completed.
    if (this.trip !== null) return;
    // Concurrent pool (opt-in): during a concurrent burst a success does NOT
    // prove the whole lane is alive (a partial rate-limit yields some
    // successes and some failures), and letting completion order decide
    // which pending victims become poison makes the trip/classification
    // non-deterministic. In this mode systemic victims stay pending — the
    // trip is count-based and order-independent; un-tripped victims end as
    // incomplete. Sequential callers omit the flag and keep the
    // poison-via-later-success attribution.
    if (this.concurrent) return;
    // The provider lane is alive: pending systemic failures were item-scoped
    // after all — poison, dead-lettered.
    for (const entry of this.pendingSystemic) this.deadLetter.push(entry);
    this.pendingSystemic = [];
  }

  /** Report an item that owes no dispatch (structural skip, budget cap, all
   * subsumed): completed for recovery-set purposes, but it proves NOTHING
   * about the provider lane — the systemic streak and pending attribution
   * are untouched. Conflating this with success would let one interleaved
   * skip reset an outage streak and write its victims off as poison. */
  recordItemSkipped(itemId: string): void {
    this.completed.push(text(itemId, "item_id"));
  }

  /** Report an item's FINAL failure (per-item budget exhausted). Returns the
   * trip state when this failure crosses the systemic threshold. */
  recordItemFailure(given: DispatchDeadLetterEntry): DispatchBreakerTripState | null {
    requireFields(given, ENTRY_FIELDS, ENTRY_FIELDS, "a failure entry");
    const entry: DispatchDeadLetterEntry = {
      item_id: text(given.item_id, "item_id"),
      failure_class:
        given.failure_class == null
          ? null
          : (text(given.failure_class, "failure_class") as SystemicDispatchFailureClass),
      failure_message: text(given.failure_message, "failure_message"),
      attempt_count: whole(given.attempt_count, "attempt_count"),
    };
    if (entry.failure_class == null) {
      // Item-local failure class: dead-letter, never breaker fuel.
      this.deadLetter.push(entry);
      return null;
    }
    if (
      !this.pendingSystemic.some((pending) => pending.item_id === entry.item_id)
    ) {
      // Post-trip in-flight systemic failures still join the pending set —
      // they are outage victims and belong to the incomplete recovery set.
      this.pendingSystemic.push(entry);
    }
    if (
      this.enabled &&
      this.trip === null &&
      this.pendingSystemic.length >= this.threshold
    ) {
      // The FIRST crossing is the trip authority; later records must not
      // rewrite its count. Concurrent-mode guarantee: the trip DECISION
      // (bool), `consecutive_item_count`, and the completed/dead-letter/
      // incomplete SETS are order-independent (pendingSystemic is never
      // flushed). `failure_class` is NOT: the trip fires early on the
      // first-N-to-complete prefix, so a mixed-class burst labels the trip by
      // whichever systemic class happened to cross. It is a best-effort
      // diagnostic label, never recovery-relevant, and is left as the
      // crossing item's class.
      this.trip = {
        ...verdict(
          "warning",
          "breaker_tripped",
          `${this.pendingSystemic.length} items in a row failed with "${entry.failure_class}",` +
            " which is the threshold: the provider, and not the items, is the likely cause",
        ),
        failure_class: entry.failure_class,
        consecutive_item_count: this.pendingSystemic.length,
        threshold: this.threshold,
      };
      return this.trip;
    }
    return null;
  }

  tripped(): DispatchBreakerTripState | null {
    return this.trip;
  }

  completedItemIds(): readonly string[] {
    return this.completed;
  }

  deadLetterEntries(): readonly DispatchDeadLetterEntry[] {
    return this.deadLetter;
  }
}

File: go/breaker.go
package haltrule

// Dispatch-loop circuit breaker - pure policy: no clock and no waiting. It
// classifies a failure message into a systemic class, computes capped
// exponential backoff delays, and tracks a batch's systemic-failure streak to
// decide trip, dead letter and completed. The loop, the retries and what is
// persisted are the caller's.

import (
	"errors"
	"fmt"
	"math"
	"strings"
	"unicode"
)

// FailureClass is what a failure message says about the provider. A nil
// *FailureClass means the failure is the item's own and says nothing.
type FailureClass string

const (
	RateLimit FailureClass = "rate_limit"
	Auth      FailureClass = "auth"
	Transport FailureClass = "transport"
)

var rateLimitPatterns = []string{
	"429", "rate limit", "limit reached", "rate_limit", "too many requests",
	"overloaded", "selected model is at capacity", "session limit",
	"usage limit", "quota", "retry-after", "retry_after",
}

var authPatterns = []string{
	"401", "403", "unauthorized", "forbidden", "invalid api key",
	"invalid x-api-key", "authentication", "not logged in",
}

// TransientTransportMessagePatterns is its own list because retry decisions
// outside this policy may want the same substrings without the rest.
var TransientTransportMessagePatterns = []string{
	"stream disconnected before completion", "connection reset by peer",
	"error sending request", "failed to connect to websocket",
	"transport channel closed", "http/request failed", "request failed after",
}

var transportPatterns = append(append([]string{}, TransientTransportMessagePatterns...),
	"timed out", "timeout", "econnrefused", "econnreset", "etimedout",
	"socket hang up", "fetch failed")

// lowerForPatterns is Unicode's full default lowercasing, which Go's per-rune
// strings.ToLower is not. The two differ in one way a pattern can see: U+0130
// lowercases to "i" followed by U+0307, so the i in it is not the i inside a
// pattern. Every other difference - the final sigma - makes a non-ASCII
// letter, and every pattern here is ASCII.
func lowerForPatterns(message string) string {
	if !strings.ContainsRune(message, 'İ') {
		return strings.ToLower(message)
	}
	var lowered strings.Builder
	for _, character := range message {
		if character == 'İ' {
			lowered.WriteString("i̇")
			continue
		}
		lowered.WriteRune(unicode.ToLower(character))
	}
	return lowered.String()
}

func holdsAny(message string, patterns []string) bool {
	for _, pattern := range patterns {
		if strings.Contains(message, pattern) {
			return true
		}
	}
	return false
}

// ClassifySystemicDispatchFailure answers the class a failure message names,
// or nil when the failure is the item's own. An absent or empty message is
// nil. The classes are tried in order, and a pattern is a plain substring.
func ClassifySystemicDispatchFailure(message string) *FailureClass {
	if message == "" {
		return nil
	}
	lowered := lowerForPatterns(message)
	for _, tried := range []struct {
		class    FailureClass
		patterns []string
	}{
		{RateLimit, rateLimitPatterns},
		{Auth, authPatterns},
		{Transport, transportPatterns},
	} {
		if holdsAny(lowered, tried.patterns) {
			class := tried.class
			return &class
		}
	}
	return nil
}

// DispatchBackoffDelayMs is the delay before retry attempt+1, with no jitter:
// capMs when initialMs is zero or less, otherwise min(capMs, initialMs * 2^max(0, attempt)).
//
// Computed in float64, as the other ports compute it: the spec's numbers are
// integers within +/-(2^53 - 1), which a float64 holds exactly, and a power of
// two times one of them is exact until it is infinite. A product past the
// range is therefore the cap and never an overflow.
func DispatchBackoffDelayMs(attempt, initialMs, capMs int64) (int64, error) {
	if err := errors.Join(
		whole(attempt, "attempt"),
		whole(initialMs, "initial_ms"),
		whole(capMs, "cap_ms"),
	); err != nil {
		return 0, err
	}
	power := math.Pow(2, math.Max(0, float64(attempt)))
	bounded := math.Min(float64(capMs), float64(initialMs)*power)
	if !math.IsInf(bounded, 0) && !math.IsNaN(bounded) && bounded > 0 {
		return int64(math.Floor(bounded)), nil
	}
	return capMs, nil
}

// wholeMax is 2^53 - 1: the largest integer every language holds, and so the
// breaker's. A Go int64 holds more, which is why the bound is written here.
const wholeMax = 1<<53 - 1

// whole refuses an argument outside the range every port shares.
func whole(value int64, what string) error {
	if value < -wholeMax || value > wholeMax {
		return fmt.Errorf("%s must be within +/-(2^53 - 1), got %d", what, value)
	}
	return nil
}

// DispatchBreakerPolicy is how one batch is judged. PerCallMaxAttempts,
// BackoffInitialMs and BackoffCapMs are carried for the caller's loop and read
// by nothing here.
type DispatchBreakerPolicy struct {
	Enabled           bool
	SystemicThreshold int64
	// Concurrent, when on, keeps a pre-trip success from flushing the pending
	// systemic failures, so which items end where does not depend on the order
	// they finished in. Off by default, which is what a sequential caller wants.
	Concurrent         bool
	PerCallMaxAttempts int64
	BackoffInitialMs   int64
	BackoffCapMs       int64
}

// DispatchDeadLetterEntry is one item's final failure, as the caller reports it.
type DispatchDeadLetterEntry struct {
	ItemID         string
	FailureClass   *FailureClass
	FailureMessage string
	AttemptCount   int64
}

// DispatchBreakerTripState is the batch's trip, once it has one.
type DispatchBreakerTripState struct {
	FailureClass         FailureClass
	ConsecutiveItemCount int64
	Threshold            int64
}

// DispatchBreakerState is the breaker over one batch. The caller reports each
// item's final outcome, after its own retries.
//
// A systemic failure is held pending until the batch proves the provider is
// alive, which a later success does: only then was it the item's own, and it
// is dead-lettered. If the streak instead reaches the threshold the batch
// trips, and the pending entries stay out of the dead letter: they are the
// outage's victims, to be dispatched again. A failure whose class is nil is
// the item's own and is dead-lettered at once.
type DispatchBreakerState struct {
	policy          DispatchBreakerPolicy
	pendingSystemic []DispatchDeadLetterEntry
	trip            *DispatchBreakerTripState
	completed       []string
	deadLetter      []DispatchDeadLetterEntry
}

// NewDispatchBreakerState starts a batch, or refuses a policy outside the
// contract - the call fails and there is no batch.
func NewDispatchBreakerState(policy DispatchBreakerPolicy) (*DispatchBreakerState, error) {
	if err := errors.Join(
		whole(policy.SystemicThreshold, "systemic_threshold"),
		// Carried for the caller's loop and read by nothing here - and in the
		// contract all the same.
		whole(policy.PerCallMaxAttempts, "per_call_max_attempts"),
		whole(policy.BackoffInitialMs, "backoff_initial_ms"),
		whole(policy.BackoffCapMs, "backoff_cap_ms"),
	); err != nil {
		return nil, err
	}
	if policy.SystemicThreshold < 1 {
		return nil, fmt.Errorf("systemic_threshold must be at least 1, got %d", policy.SystemicThreshold)
	}
	return &DispatchBreakerState{policy: policy}, nil
}

// RecordItemSuccess reports a real dispatch success, the only event that
// proves the provider is alive. An item that made no call uses RecordItemSkipped.
func (state *DispatchBreakerState) RecordItemSuccess(itemID string) {
	state.completed = append(state.completed, itemID)
	// Attribution freezes at the trip: a late success must not write the
	// outage's victims off as the items' own failures.
	if state.trip != nil {
		return
	}
	if state.policy.Concurrent {
		return
	}
	state.deadLetter = append(state.deadLetter, state.pendingSystemic...)
	state.pendingSystemic = nil
}

// RecordItemSkipped reports an item that owed no dispatch. It is completed,
// and it proves nothing about the provider.
func (state *DispatchBreakerState) RecordItemSkipped(itemID string) {
	state.completed = append(state.completed, itemID)
}

// RecordItemFailure reports an item's final failure. It answers the trip when
// this failure is the one that crosses the threshold, and nil otherwise.
func (state *DispatchBreakerState) RecordItemFailure(entry DispatchDeadLetterEntry) (*DispatchBreakerTripState, error) {
	if err := whole(entry.AttemptCount, "attempt_count"); err != nil {
		return nil, err
	}
	if entry.FailureClass == nil {
		state.deadLetter = append(state.deadLetter, entry)
		return nil, nil
	}
	pending := false
	for _, held := range state.pendingSystemic {
		if held.ItemID == entry.ItemID {
			pending = true
			break
		}
	}
	if !pending {
		// A post-trip systemic failure still joins the pending entries: it is
		// an outage victim and belongs to the incomplete set.
		state.pendingSystemic = append(state.pendingSystemic, entry)
	}
	if state.policy.Enabled && state.trip == nil &&
		int64(len(state.pendingSystemic)) >= state.policy.SystemicThreshold {
		// The first crossing is the trip, and later reports do not rewrite it.
		state.trip = &DispatchBreakerTripState{
			FailureClass:         *entry.FailureClass,
			ConsecutiveItemCount: int64(len(state.pendingSystemic)),
			Threshold:            state.policy.SystemicThreshold,
		}
		return state.trip, nil
	}
	return nil, nil
}

// Tripped is the batch's trip, or nil while it has none.
func (state *DispatchBreakerState) Tripped() *DispatchBreakerTripState {
	return state.trip
}

// CompletedItemIDs are the ids reported complete, in the order they were
// reported, success and skipped alike.
func (state *DispatchBreakerState) CompletedItemIDs() []string {
	return state.completed
}

// DeadLetterEntries are the entries written off, in the order they arrived.
func (state *DispatchBreakerState) DeadLetterEntries() []DispatchDeadLetterEntry {
	return state.deadLetter
}
